libbech32/bech32.cpp

```diff
@@ -314,50 +347,132 @@ const char * bech32_strerror(bech32_error error_code) {
+    hrpdp->hrp = static_cast<char *>(calloc(hrpdp->hrplen+1, 1)); // +1 for '\0'
     if(hrpdp->hrp == nullptr) {
         free(hrpdp);
         return nullptr;
     }
-    hrpdp->dplen = number_of_data_characters + 1; // +1 for '\0'
+    hrpdp->dplen = number_of_data_characters;
     hrpdp->dp = static_cast<unsigned char *>(calloc(hrpdp->dplen, 1));
     if(hrpdp->dp == nullptr) {
         free(hrpdp->hrp);
         free(hrpdp);
         return nullptr;
     }
 
+    hrpdp->encoding = ENCODING_UNKNOWN;
+
     return hrpdp;
 }
 
 /**
- * Frees memory for a txref_LocationData struct.
+ * Frees memory for a bech32_HrpAndDp struct.
+ *
+ * @param hrpAndDp a pointer to a bech32_HrpAndDp struct
  */
 extern "C"
-void free_HrpAndDp_storage(bech32_HrpAndDp *hrpdp) {
-    free(hrpdp->dp);
-    free(hrpdp->hrp);
-    free(hrpdp);
+void bech32_free_HrpAndDp(bech32_HrpAndDp *hrpAndDp) {
+    if(hrpAndDp == nullptr)
+        return;
+    free(hrpAndDp->dp);
+    free(hrpAndDp->hrp);
+    free(hrpAndDp);
+}
+
+/**
+ * Computes final length for a to-be-encoded bech32 string
+ *
+ * @param hrplen the length of the "human-readable part" string. must be > 0
+ * @param dplen the length of the "data part" array
+ *
+ * @return length of to-be-encoded bech32 string
+ */
+extern "C"
+size_t bech32_compute_encoded_string_length(size_t hrplen, size_t dplen) {
+    return hrplen + SEPARATOR_LENGTH + dplen + CHECKSUM_LENGTH;
+}
+
+/**
+ * Allocates memory for a to-be-encoded bech32 string
+ *
+ * This memory must be freed using the bech32_free_bstring function.
+ *
+ * @param hrplen the length of the "human-readable part" string. must be > 0
+ * @param dplen the length of the "data part" array
+ *
+ * @return a pointer to a new bech32_bstring struct, or NULL if error
+ */
+extern "C"
+bech32_bstring * bech32_create_bstring(size_t hrplen, size_t dplen) {
+    if(hrplen < 1)
+        return nullptr;
+    auto *bstring = static_cast<bech32_bstring *>(malloc(sizeof(bech32_bstring)));
+    bstring->length = bech32_compute_encoded_string_length(hrplen, dplen);
+    bstring->string = static_cast<char *>(calloc(bstring->length + 1, 1)); // +1 for '\0'
+    return bstring;
+}
+
+/**
+ * Allocates memory for a to-be-encoded bech32 string based on the size of the bech32_HrpAndDp struct
+ *
+ * This memory must be freed using the bech32_free_bstring function.
+ *
+ * @param hrpAndDp a pointer to a bech32_HrpAndDp struct
+ *
+ * @return a pointer to a new bech32_bstring struct, or NULL if error
+ */
+extern "C"
+bech32_bstring * bech32_create_bstring_from_HrpAndDp(bech32_HrpAndDp *hrpAndDp) {
+    if(hrpAndDp == nullptr)
+        return nullptr;
+    if(hrpAndDp->hrplen < 1)
+        return nullptr;
+    return bech32_create_bstring(hrpAndDp->hrplen, hrpAndDp->dplen);
+}
+
+/**
+ * Frees memory for a bech32 string.
+ *
+ * @param bstring pointer to a bech32 string
+ */
+extern "C"
+void bech32_free_bstring(bech32_bstring *bstring) {
+    if(bstring == nullptr)
+        return;
```

Comment:
same Q here: is the program working well if it frees a null pointer?
discussion:
  https://stackoverflow.com/questions/4190703/is-it-safe-to-delete-a-null-pointer#comment34158001_4190737